### PR TITLE
refactor generator

### PR DESCRIPTION
**Pull request by ccoVeille**
- **gen: move assertion sorter to a function so it could be reused**
- **bool_compare: refactor tests**

These commits were part of ~#107~ and #111 that were suggested changes were discarded, but the refactoring made in each PR could be kept as discussed [here](https://github.com/Antonboom/testifylint/issues/110#issuecomment-2155469152)

### CODE REVIEW

**Comment by ccoVeille** [comment]
This PR uses commits available in:
- #116 created for #106
- #111 created for #110

**Comment by ccoVeille** [comment]
@Antonboom do you think we could merge this one, it would help me to refactor for 
- #94

File: internal/testgen/main.go [hunk 1]
```diff
@@ -135,3 +136,13 @@ func genGoFileFromTmpl(output string, exec Executor, data any) error {
 
 	return os.WriteFile(output, formatted, 0o644)
 }
+
+func sortAssertions(assertions []Assertion) {
```
Comment (Antonboom) [inline internal/testgen/main.go:140]: move it to `assertion.go`

let's keep `main.go` free from any helpers, the file is about generator itself

File: internal/testgen/gen_bool_compare.go [hunk 2]
```diff
@@ -26,6 +26,140 @@ func (g BoolCompareTestsGenerator) TemplateData() any {
 		ValidAssertions   []Assertion
 	}
 
+	ignoredAssertions := make([]Assertion, 0)
```
Comment (Antonboom) [inline internal/testgen/gen_bool_compare.go:29]: https://github.com/uber-go/guide/blob/master/style.md#nil-is-a-valid-slice

or write
```go
var ignoredAssertions []Assertion
```

or use real capacity
```go
ignoredAssertions := make([]Assertion, 2*8+6)
// but! magic numbers + no make sense for this case anyway
```

File: internal/testgen/gen_bool_compare.go [hunk 3]
```diff
@@ -26,6 +26,140 @@ func (g BoolCompareTestsGenerator) TemplateData() any {
...
+	ignoredAssertions = append(ignoredAssertions,
+		Assertion{Fn: "Equal", Argsf: "foo, foo"},
+		Assertion{Fn: "NotEqual", Argsf: "foo, foo"},
+		Assertion{Fn: "True", Argsf: "foo == foo"},
+		Assertion{Fn: "False", Argsf: "foo == foo"},
+		Assertion{Fn: "True", Argsf: "foo != foo"},
+		Assertion{Fn: "False", Argsf: "foo != foo"},
+	)
+
+	invalidAssertionsForTrue := make([]Assertion, 0)
```
Comment (Antonboom) [inline internal/testgen/gen_bool_compare.go:87]: var

File: internal/testgen/gen_bool_compare.go [hunk 4]
```diff
@@ -26,6 +26,140 @@ func (g BoolCompareTestsGenerator) TemplateData() any {
...
+	invalidAssertionsForFalse := make([]Assertion, 0)
```
Comment (Antonboom) [inline internal/testgen/gen_bool_compare.go:88]: var

File: internal/testgen/gen_compares.go [hunk 5]
```diff
@@ -22,41 +22,52 @@ func (g ComparesTestsGenerator) TemplateData() any {
...
+	}
+
+	invalidAssertions = append(invalidAssertions,
+		Assertion{Fn: "True", Argsf: "ptrA == ptrB", ReportMsgf: report, ProposedFn: "Same", ProposedArgsf: "ptrA, ptrB"},
+		Assertion{Fn: "True", Argsf: "ptrA != ptrB", ReportMsgf: report, ProposedFn: "NotSame", ProposedArgsf: "ptrA, ptrB"},
+		Assertion{Fn: "False", Argsf: "ptrA == ptrB", ReportMsgf: report, ProposedFn: "NotSame", ProposedArgsf: "ptrA, ptrB"},
+		Assertion{Fn: "False", Argsf: "ptrA != ptrB", ReportMsgf: report, ProposedFn: "Same", ProposedArgsf: "ptrA, ptrB"},
+	)
+
+	sortAssertions(invalidAssertions)
```
Comment (Antonboom) [inline internal/testgen/gen_compares.go:60]: any reason to sort it?
the order of assertions is determined

File: internal/testgen/gen_useless_assert.go [hunk 6]
```diff
@@ -71,13 +70,7 @@ func (g UselessAssertTestsGenerator) TemplateData() any {
...
 	}
 
-	sort.Slice(twoSideAssertions, func(i, j int) bool {
-		lhs, rhs := twoSideAssertions[i], twoSideAssertions[j]
-		if lhs.Fn == rhs.Fn {
-			return lhs.Argsf < rhs.Argsf
-		}
-		return lhs.Fn < rhs.Fn
-	})
+	sortAssertions(twoSideAssertions)
```
Comment (Antonboom) [inline internal/testgen/gen_useless_assert.go:73]: let's look we can just leave the sort code here only

because there is `map` above

File: internal/testgen/gen_bool_compare.go [hunk 7]
```diff
@@ -26,6 +26,140 @@ func (g BoolCompareTestsGenerator) TemplateData() any {
...
+		invalidAssertionsForTrue = append(invalidAssertionsForTrue,
+			Assertion{Fn: "True", Argsf: argsf, ReportMsgf: reportSimplify, ProposedArgsf: "predicate"},
+		)
+
+		invalidAssertionsForFalse = append(invalidAssertionsForFalse,
+			Assertion{Fn: "False", Argsf: argsf, ReportMsgf: reportSimplify, ProposedArgsf: "predicate"},
+		)
+	}
+
+	sortAssertions(invalidAssertionsForTrue)
```
Comment (Antonboom) [inline internal/testgen/gen_bool_compare.go:159]: any reason to sort it?
the order of assertions is determined

safe this PR from unrelated changes around generated test files

File: internal/testgen/gen_compares.go [hunk 8]
```diff
@@ -22,41 +22,52 @@ func (g ComparesTestsGenerator) TemplateData() any {
...
-		ignored = append(ignored,
+		ignoredAssertions = append(ignoredAssertions,
 			Assertion{Fn: "True", Argsf: fmt.Sprintf("c %s d", tok)},
 			Assertion{Fn: "False", Argsf: fmt.Sprintf("d %s c", tok)},
 		)
 	}
 
+	var invalidAssertions []Assertion //nolint:prealloc
+	for checker, oppositeChecker := range map[string]string{
+		"True": "False",
```
Comment (Antonboom) [inline internal/testgen/gen_compares.go:35]: loop and map doesn't make sense in such case
this is only increase complexity of support

please revert it or do in the same way as `append` below

**Review by Antonboom — CHANGES_REQUESTED** [review]

File: internal/testgen/main.go (hunk 1 above)
Comment (ccoVeille) [reply internal/testgen/main.go]: Sure. Thanks for the review

**Comment by ccoVeille** [comment]
I pushed too fast sorry.

I will fix other too

File: internal/testgen/gen_compares.go (hunk 8 above)
Comment (ccoVeille) [reply internal/testgen/gen_compares.go:35]: This one was there, because #111 had another commit to add things 1eb5934a8f43d09422d826ac93987ae2e40e882d

I will revert sure

File: internal/testgen/gen_useless_assert.go (hunk 6 above)
Comment (ccoVeille) [reply internal/testgen/gen_useless_assert.go:73]: OK, I will remove the others

File: internal/testgen/gen_bool_compare.go (hunk 7 above)
Comment (ccoVeille) [reply internal/testgen/gen_bool_compare.go:159]: I will remove it then

**Comment by ccoVeille** [comment]
I resolved all your comments @Antonboom please check and review.

The number of changes is then very limited
